### ISSUE TITLE
CURP - fast replication nearly for free and without the current caveats

### ISSUE THREAD

**Issue by dumblob** [JohnSully/KeyDB#49]
There is a "new" approach to very fast replication called [CURP](https://blog.acolyer.org/2019/03/15/exploiting-commutativity-for-practical-fast-replication/ ) (Consistent Unordered Replication Protocol).

It shall not have the caveats the current replication scheme in Redis has and shall not be difficult to implement (actually it was already partially implemented for Redis to test it - don't know though where the source lies and what is the license).

Would you consider implementing such low hanging fruit?

I've made a [similar ticket directly for Redis](https://github.com/antirez/redis/issues/6175 ), but as Redis is under the hood of a company with specific opinions, I'm opening it here as well as I'd like Redis-like DBs thrive.

**Comment by JohnSully** [comment]
It looks very interesting. Yes i will be investigating this.

**Comment by dumblob** [comment]
Did you have some time to take a look at that paper?

**Comment by JohnSully** [comment]
@dumblob Yes I have looked, but its not quite time to implement this for KeyDB just yet. 

Right now the focus for KeyDB is maximizing vertical scaling. In support of that we have MVCC coming out this week. Once we’ve maximized what a single node can do I will start to take a deeper look into CDRT/CURPS for active replication.

**Comment by dumblob** [comment]
@JohnSully understood, thanks for the update. We're considering switching to cloud based international provider of Redis API (does KeyDB.dev or EQ ALPHA provide such services?), but with way better scalability (including failover/HA), throughput and pushed limits (e.g. limitation by RAM) while guaranteeing **full durability** (we currently use `appendonly yes` and `appendfsync always`).

Yet another requirement on the chosen technology is the possibility to run everything as an on-premise DB (without any restrictions) for some special customers (in other words the chosen technology has to be open source without any limitations for commercial use).
